### ISSUE TITLE
x/tools/cmd/vet: structtag field repeats json warning with valid override

### ISSUE THREAD

**Issue by tylfin** [golang/go#40102]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.14.1 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/[XX]/Library/Caches/go-build"
GOENV="/Users/[XX]/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/[XX]/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.14.1/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.14.1/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/w9/kv18_grs18x65fyzv1p5mvk00000gn/T/go-build893878327=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

When embedding two structs with duplicate JSON tags, e.g.:

```
type A struct {
   name string `json:"name"`
   // additional fields
}

type B struct {
   name string `json:"name"`
   ... // additional fields
}
```

A reasonable workaround to the `go vet` structtag warning is to override the duplicate-tag on the top level, per extra rules here: https://golang.org/pkg/encoding/json/#Marshal, e.g.

```
type C struct {
   A
   B
   name string `json:"name"`
}
```

But `go vet` still raises the warning.

### What did you expect to see?

`go vet` treats top-level `structtag` overrides as a valid solution and no longer raises a warning.

### What did you see instead?

`go vet` ignores the top-level override, resulting in `structtag: struct field X repeats json tag "[X]" also at XXXX.go:XXX (govet)`

**Comment by vedranvuk** [comment]
It does this for the "-" tag as well.

**Comment by ahmednawaz10p** [comment]
any update on this ?

**Comment by sc0Vu** [comment]
It happened to me. 

Is that possible to skip json package here: https://cs.opensource.google/go/x/tools/+/refs/tags/v0.40.0:go/analysis/passes/structtag/structtag.go;l=203?

**Comment by adonovan** [comment]
(Sorry for the delay.)

Is this actually a bug? Consider this program based on the original example: https://go.dev/play/p/DS2pPZEQX0N.

```go
type A struct {
	Name string `json:"name"`
}

type B struct {
	Name string `json:"name2"`
}

type C struct {
	A
	B  // "struct field Name repeats json tag "name" also at a.go:4"
}

func main() {
	var c C
	c.A.Name = "a"
	c.B.Name = "b"
	data, _ := json.Marshal(c)
	fmt.Println(string(data))
}
```

The name and name2 fields are siblings in the printed JSON output, `{"name":"a","name2":"b"}`. If you now rename the `name2` tag to `name`, the JSON output becomes just `{}`, because of the name conflict, and the vet check reports "struct field Name repeats json tag "name" also at a.go:4". So that seems to be working  as intended.

Now, your proposed workaround is to add a tag on the C.B field, such as `json:"name3"`.
```go
type C struct {
	A
	B `json:"name3"` // "struct field Name repeats json tag "name" also at a.go:4"
}
```
This has the IMHO surprising effect  of changing the structure of the JSON output so that the fields of A and B are no longer siblings, but B is an explicit object: `{"name":"a","name3":{"name2":"b"}}`. The vet check is silent in this case, which is I think exactly what was being asked for in this issue.

In short, I don't think the change proposed in this issue makes sense, but perhaps I am confused.

**Comment by tylfin** [comment]
Unfortunately, I don't have any recollection of the context here 🤷 , but you can probably ignore my suggested solution from back then.

EDIT: Actually, IIRC this was a closer example to what I was running into, and I wanted an inline comment to silence `go vet` https://go.dev/play/p/KsVC698KLjr
